### PR TITLE
jest-sorted

### PR DESCRIPTION
**Pull request by baloglu**
This PR:
 - Adds an /api/articles endpoint that returns array of articles sorted by created_at field
 - Add tests for this endpoint

### CODE REVIEW

File: __tests__/api.test.js [hunk 1]
```diff
@@ -63,4 +63,32 @@ describe("API endpoints", () => {
...
         });
     })
+    test("/api/articles endpoint should return all articles in an array", () => {
+        return request(app)
+        .get("/api/articles")
+        .expect(200)
+            .then((result) => {
+                const articles = result.body.articles;
+                const expected = 12
+                expect(articles.length).toBe(expected)
```
Comment (leekli) [inline __tests__/api.test.js:73]: Good assertions here Chetin, well done. Good to check the length, and checking that structure of the data (the shape) is as you expect it to be.

File: __tests__/api.test.js [hunk 2]
```diff
@@ -63,4 +63,32 @@ describe("API endpoints", () => {
...
+        });
+    })
+    test("/api/articles should return array of articles in date descending order", () => {
+        return request(app)
+        .get("/api/articles")
+        .expect(200)
+            .then((result) => {
+                const articles = result.body.articles;
+                const expected = 'created_at'
+                expect(articles).toBeSortedBy(expected, {descending: true})
```
Comment (leekli) [inline __tests__/api.test.js:91]: Excellent, good to see this test to check that articles are sorted by their date, well done.

File: controllers/get-articles.controller.js [hunk 3]
```diff
@@ -15,4 +15,13 @@ function getArticlesId(request, response, next) {
     
 }
 
-module.exports = getArticlesId
\ No newline at end of file
+function getArticles(request, response, next) {
+    getArticlesModel()
+    .then(result => {
+        return response.status(200).send({ articles: result.rows } );
```
Comment (leekli) [inline controllers/get-articles.controller.js]: You shouldn't need the return key word here.

Also, sending the response on a key of 'articles' to represent that it's articles (plural) which are being returned to the client is good, and a great use of names - well done

File: controllers/get-articles.controller.js [hunk 4]
```diff
@@ -15,4 +15,13 @@ function getArticlesId(request, response, next) {
...
+        })
+        .catch(err => {
```
Comment (leekli) [inline controllers/get-articles.controller.js:20]: Good to see a catch block here to deal with any errors :)

File: models/get-articles.model.js [hunk 5]
```diff
@@ -15,4 +15,24 @@ function getArticleIdModel(article_id) {
...
 }
 
-module.exports = getArticleIdModel
\ No newline at end of file
+function getArticlesModel() {
+    // comment_count which is the total count of all 
+    // the comments with this article_id - you should make
+    // use of queries to the database in order to achieve this.
+    
+    //  the articles should be sorted by date in descending order.
```
Comment (leekli) [inline models/get-articles.model.js]: It's worth removing these comments now

File: models/get-articles.model.js [hunk 6]
```diff
@@ -15,4 +15,24 @@ function getArticleIdModel(article_id) {
...
+     FROM articles LEFT JOIN comments ON articles.article_id = comments.article_id ORDER BY articles.created_at DESC;`)
+        .then(result => {
+            if (result.rows.length === 0) {
+                return Promise.reject({ status: 400, msg: "There are no articles" })
+            }
+            return result;
+        })
+        .catch(err => {
+            return Promise.reject(err)
+        });
```
Comment (leekli) [inline models/get-articles.model.js]: Your model doesn't need a .catch block, any errors should be caught in the controller, so please do remove this catch block

File: models/get-articles.model.js [hunk 7]
```diff
@@ -15,4 +15,24 @@ function getArticleIdModel(article_id) {
...
\ No newline at end of file
+function getArticlesModel() {
+    // comment_count which is the total count of all 
+    // the comments with this article_id - you should make
+    // use of queries to the database in order to achieve this.
+    
+    //  the articles should be sorted by date in descending order.
+    return db.query(`SELECT DISTINCT articles.article_id, articles.author, articles.title, articles.topic, articles.created_at, 
+    articles.votes, articles.article_img_url, (SELECT count(*) from comments WHERE comments.article_id = articles.article_id) as comment_count
+     FROM articles LEFT JOIN comments ON articles.article_id = comments.article_id ORDER BY articles.created_at DESC;`)
```
Comment (leekli) [inline models/get-articles.model.js:22]: Good work here with using an aggregate function (COUNT) to do the counting of total comments for you, well done!

**Review by leekli — APPROVED** [review]
Well done Chetin, I have left comments on this pull request.

You are free to merge this branch, I've made recommendations for some slight amendments for you to make prior to merging through, but you don't to submit another pull request for that - I'm happy for you to make the changes, commit them and then merge the pull request

**Comment by baloglu** [comment]
> Well done Chetin, I have left comments on this pull request.
> 
> You are free to merge this branch, I've made recommendations for some slight amendments for you to make prior to merging through, but you don't to submit another pull request for that - I'm happy for you to make the changes, commit them and then merge the pull request

Thank you for all the comments Lee, I applied them and merging this PR now!

**Pull request merged**